### ISSUE TITLE
Something strange?

### ISSUE THREAD

**Issue by yypurpose** [Majiker/BalancedMetaSoftmax-InstanceSeg#5]
Hi! I was reading your code and found something strange. In Ln. [151](https://github.com/Majiker/BalancedMetaSoftmax-InstanceSeg/blob/b7d4bd9c71c583cf3ad9e349e2e1a693b5ad4012/detectron2/modeling/roi_heads/fast_rcnn.py#L151) of fast_rcnn.py, pred_class_logits includes the background class. While the weight in Ln. [71](https://github.com/Majiker/BalancedMetaSoftmax-InstanceSeg/blob/main/projects/BALMS/balms/fast_rcnn.py#L71) in balms/fast_rcnn.py dose not. So, in the formula **self.pred_class_logits[fg_ind] = (self.pred_class_logits - weight.log())[fg_ind]**, there is a size mismatch, isn't it? self.pred_class_logits is (R, K+1) while weight.log() is (R, K).

This part causes some confusion for me. Hope you could help me.

Thanks in advance.

**Comment by jiawei-ren** [comment]
There is no size mismatch. `pred_class_logits` does _not_ include the background class in the BALMS project. Its size is (R, K) instead of (R, K+1). 

To use multiple sigmoid regression, we removed the prediction on the backgorund class by modifying Detectron's RCNN head [here](https://github.com/Majiker/BalancedMetaSoftmax-InstanceSeg/blob/main/projects/BALMS/balms/fast_rcnn.py#L126). The prediction size in BALMS should be `num_class` instead of the original `num_class+1`, which matches the size of `weight`.

**Comment by yypurpose** [comment]
Ooooh! thanks a lot!

**Comment by yypurpose** [comment]
And so where is the loss for background samples, as the paper said "applying the standard Sigmoid function to background samples". THANKS A LOT!!

**Comment by jiawei-ren** [comment]
> Where is the loss for background samples

We compute the loss for all samples [here](https://github.com/Majiker/BalancedMetaSoftmax-InstanceSeg/blob/main/projects/BALMS/balms/fast_rcnn.py#L73), where the loss for background samples is also included.

> as the paper said "applying the standard Sigmoid function to background samples"

This is to describe what `fg_ind` does in Ln. [71](https://github.com/Majiker/BalancedMetaSoftmax-InstanceSeg/blob/main/projects/BALMS/balms/fast_rcnn.py#L71) in balms/fast_rcnn.py. No weights are applied to background samples.

**Comment by yypurpose** [comment]
I got it, thank u

**Comment by jiawei-ren** [comment]
I will close the issue for now. Feel free to reopen it if you have any further questions.

**Comment by yypurpose** [comment]
Hi, is your changes to the original sigmoid loss just removing the background category and applying balanced softmax on foreground samples? Cause recently, I'm trying to apply your work using mmdetection, but found the performance drops a lot once remove the background category.

I still don't know how to fix it, so I turn to you for help again ˃̣̣̥᷄⌓˂̣̣̥᷅

Thanks a lot!

**Comment by jiawei-ren** [comment]
> is your changes to the original sigmoid loss just removing the background category and applying balanced softmax on foreground samples?

Yes, and it should be balanced sigmoid to be more exact.

> Cause recently, I'm trying to apply your work using mmdetection, but found the performance drops a lot once remove the background category.

I actually have a plan to create a PR to MMDet in the following weeks (on the loss part only). Is your repo publicly available? I may be able to help to review the code if I have any free time.

**Comment by yypurpose** [comment]
Yes! I made my repo available just a few moment ago [here](https://github.com/yypurpose/mmdetection) (with ugly code...). 

I trained mask-rcnn on LVIS_v1.0, and used the [this](https://github.com/yypurpose/mmdetection/blob/main/configs/det_lt/mask_rcnn_50_balancedsoftmax_lvis_sigmoid.py) config.

The main modification is in [L80](https://github.com/yypurpose/mmdetection/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py#L80) and [L137](https://github.com/yypurpose/mmdetection/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py#L137) of convfc_bbox_head.py. I naively changed the num_class of cls_fc and the initialization of bias. And added BalancedSoftmaxCrossEntropyLoss (or balanced sigmoid) in [cross_entropy_loss.py](https://github.com/yypurpose/mmdetection/blob/main/mmdet/models/losses/cross_entropy_loss.py#L219). The [self.class_weight](https://github.com/yypurpose/mmdetection/blob/main/mmdet/models/losses/cross_entropy_loss.py#L249) was computed from LVIS_v1.0.

I also clipped the gradient.

Thanks in advance!

**Comment by jiawei-ren** [comment]
Your code is using Softmax for testing. This [line](https://github.com/yypurpose/mmdetection/blob/main/mmdet/models/roi_heads/bbox_heads/bbox_head.py#L280) should be changed to something like this:
```python
        if self.use_sigmoid_cls:
            scores = torch.sigmoid(cls_score) if cls_score is not None else None
            # add a dummy probs, representing a fake score for background class
            n = scores.size(0)
            dummy_probs = scores.new_zeros(n, 1)
            scores = torch.cat([scores, dummy_probs], dim=1)
        else:
            scores = F.softmax(cls_score, dim=1) if cls_score is not None else None
```
